proposals/NNNN-outofproc-compiler-api-architecture.md

```diff
@@ -0,0 +1,408 @@
+<!-- {% raw %} -->
+
+# Out-of-process compiler api architecture
+
+* Proposal: [NNNN](NNNN-outofproc-compiler-api-architecture.md)
+* Author(s): [Cooper Partin](https://github.com/coopp)
+* Sponsor: [Cooper Partin](https://github.com/coopp)
+* Status: **Under Consideration**
+* Impacted Project(s): (Clang)
+
+* Issues:
+
+## Introduction
+
+An effort is underway to bring compilation support into clang for HLSL based
+shaders.  A C style api will be built to enable applications to compile
+a shader from their own processes in addition to being able to launching the
+clang process.
+
+This document is to propose a more detailed architecture for the approved 
+proposal: [0005](0005-inproc-outofproc-compiler-api-support.md).
+
+The c api [0008](0008-c-interface-compiler-library.md) for compiling a shader
+will be implemented using this out of process architecture.
+
+## Proposed solution
+
+The architecture for an out of process design will take advantage of systems
+that have multiple processors, or multiple-core processors. A process pool will
+be created that creates a separate compiler process for each available
+processor. For example, if the system has four processors, then four compiler
+processes are created.
+
+The process pool is owned by a singleton object and shared between each api
+instance created in the calling process.
+
+A set of low-level apis will be designed that will drive compilation work
+across multiple processes.
```

Comment:
> A set of low-level APIs will be designed that will drive compilation work across multiple processes.

IMO we should avoid saying things like "will be designed" and instead have the proposal be more definitive. However, I had trouble making a small change to make it that way. 

Since we've discussed this offline I know what this is intended to say, but I'm not convinced that is clear here.  For example, taking that sentence at face value leads to the question: "How does designing a set of low-level APIs drive compilation work across multiple processes?" And that's not what's intended here.

I think my overall suggestion here is that the low/high-level split of the libraries is a pretty fundamental part of your proposal and you should probably start the Proposed Solution section with an explanation of this. 

I don't think you're intending this proposal to be the design of the low-level API, so I think you should just have a hand-wavy description of it and a note that the design on this is out-of-scope of this proposal (in fact, the design should be guided by the requirements discovered while designing the high-level API).

After that you can go into more detail about the high-level API.

Reply:
The high-level/low-level api thing was really confusing. I removed this content.